File: app/Models/Transfer.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class Transfer extends Model
{
    use HasFactory, SoftDeletes;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'reference_number',
        'user_id',
        'driver_id',
        'driver_confirmation_status',
        'job_status',
        'job_started_at',
        'job_completed_at',
        'car_id',
        'pickup_location_id',
        'dropoff_location_id',
        'pickup_latitude',
        'pickup_longitude',
        'pickup_location_name',
        'dropoff_latitude',
        'dropoff_longitude',
        'dropoff_location_name',
        'pickup_datetime',
        'return_datetime',
        'flight_number',
        'airline',
        'passenger_count',
        'luggage_count',
        'status',
        'price',
        'currency',
        'payment_status',
        'payment_method',
        'payment_id',
        'special_instructions',
        'driver_notes',
        'cancellation_reason',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $appends = ['status_label', 'status_badge_class'];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'pickup_datetime' => 'datetime',
        'return_datetime' => 'datetime',
        'job_started_at' => 'datetime',
        'job_completed_at' => 'datetime',
        'price' => 'decimal:2',
        'pickup_latitude' => 'float',
        'pickup_longitude' => 'float',
        'dropoff_latitude' => 'float',
        'dropoff_longitude' => 'float',
    ];

    /**
     * Get the user who booked the transfer.
     */
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get the driver assigned to the transfer.
     */
    public function driver()
    {
        return $this->belongsTo(User::class, 'driver_id');
    }

    /**
     * Get the car used for the transfer.
     */
    public function car()
    {
        return $this->belongsTo(Car::class);
    }

    /**
     * Pickup and dropoff destination relationships.
     */
    public function pickupDestination()
    {
        return $this->belongsTo(Destination::class, 'pickup_location_id');
    }

    public function dropoffDestination()
    {
        return $this->belongsTo(Destination::class, 'dropoff_location_id');
    }

    /**
     * Get the human-readable status label.
     *
     * @return string
     */
    public function getStatusLabelAttribute()
    {
        $statuses = [
            'pending' => 'En attente',
            'confirmed' => 'Confirmé',
            'assigned' => 'Assigné',
            'on_the_way' => 'En route',
            'driver_en_route' => 'En route',
            'driver_arrived' => 'Chauffeur arrivé',
            'in_progress' => 'En cours',
            'completed' => 'Terminé',
            'cancelled' => 'Annulé',
            'no_show' => 'Non-présentation',
        ];

        return $statuses[$this->status] ?? ucfirst($this->status);
    }

    /**
     * Get the Bootstrap badge class for the status.
     *
     * @return string
     */
    public function getStatusBadgeClassAttribute()
    {
        $classes = [
            'pending' => 'secondary',
            'confirmed' => 'primary',
            'assigned' => 'info',
            'on_the_way' => 'warning',
            'driver_en_route' => 'warning',
            'driver_arrived' => 'info',
            'in_progress' => 'primary',
            'completed' => 'success',
            'cancelled' => 'danger',
            'no_show' => 'dark',
        ];

        return $classes[$this->status] ?? 'light';
    }

}

File: ios_app_car/lib/models/transfer.dart
class Transfer {
  final int id;
  final String pickupLocation;
  final String dropoffLocation;
  final String pickupTime;
  final String status;
  final String driverConfirmationStatus;
  final String jobStatus;
  final String? jobStartedAt;
  final String? jobCompletedAt;
  final double? pickupLatitude;
  final double? pickupLongitude;
  final double? dropoffLatitude;
  final double? dropoffLongitude;
  final User? user;
  final Car? car;

  Transfer({
    required this.id,
    required this.pickupLocation,
    required this.dropoffLocation,
    required this.pickupTime,
    required this.status,
    required this.driverConfirmationStatus,
    required this.jobStatus,
    this.jobStartedAt,
    this.jobCompletedAt,
    this.pickupLatitude,
    this.pickupLongitude,
    this.dropoffLatitude,
    this.dropoffLongitude,
    this.user,
    this.car,
  });

  static double? _parseCoordinate(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      try {
        return double.parse(value);
      } catch (e) {
        return null;
      }
    }
    return null;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    if (value is String) {
      try {
        return int.parse(value);
      } catch (e) {
        return 0;
      }
    }
    return 0;
  }

  factory Transfer.fromJson(Map<String, dynamic> json) {
    // Get location names from database fields or fallback
    String pickupLocationName = json['pickup_location_name'] ?? 
                               json['pickup_destination']?['name'] ?? 
                               'Pickup Location';
    
    String dropoffLocationName = json['dropoff_location_name'] ?? 
                                json['dropoff_destination']?['name'] ?? 
                                'Dropoff Location';

    return Transfer(
      id: _parseInt(json['id']),
      pickupLocation: pickupLocationName,
      dropoffLocation: dropoffLocationName,
      pickupTime: json['pickup_datetime']?.toString() ?? json['pickup_time']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      driverConfirmationStatus: json['driver_confirmation_status']?.toString() ?? 'pending',
      jobStatus: json['job_status']?.toString() ?? 'pending',
      jobStartedAt: json['job_started_at']?.toString(),
      jobCompletedAt: json['job_completed_at']?.toString(),
      pickupLatitude: _parseCoordinate(json['pickup_latitude']),
      pickupLongitude: _parseCoordinate(json['pickup_longitude']),
      dropoffLatitude: _parseCoordinate(json['dropoff_latitude']),
      dropoffLongitude: _parseCoordinate(json['dropoff_longitude']),
      user: json['user'] != null ? User.fromJson(json['user']) : null,
      car: json['car'] != null ? Car.fromJson(json['car']) : null,
    );
  }
}

class User {
  final int id;
  final String name;
  final String email;

  User({
    required this.id,
    required this.name,
    required this.email,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: Transfer._parseInt(json['id']),
      name: json['name']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
    );
  }
}

class Car {
  final int id;
  final String brand;
  final String model;
  final String? image;

  Car({
    required this.id,
    required this.brand,
    required this.model,
    this.image,
  });

  factory Car.fromJson(Map<String, dynamic> json) {
    return Car(
      id: Transfer._parseInt(json['id']),
      brand: json['brand']?.toString() ?? '',
      model: json['model']?.toString() ?? '',
      image: json['image']?.toString(),
    );
  }
}
